Rev2: types and functions

achaikou opened the pull request:
- get/set field and field_forall are supported
- floats and strings (for header names) are supported
- all data is available in python, but python's user interface is not updated yet.

achaikou committed:
Add string type support

Note that type itself is not changed in the main header per request from
PO. We still consider those two values to be Unassigned 1 and Unassigned
2 and not "SEG00000"

We leave these values as is in case someone used them to store optional
values.

In rev 1.0
> 233-240 Unassigned — For optional information.

achaikou committed:
Make getfield/putfield methods on segyfd

That might cause breaking changes: before users could use this function
even when file was closed. No longer.
We don't expect this to be an issue.

Alternative solution would be passing trace header mapping along with
the header, but that would require mapping re-parsing.

achaikou committed:
Set endianness before extended text headers parse

Stanzas were gathered incorrectly for little endian files as endianness
was still assumed to be big-endian.

There is a bit of a design problem.

Operations flow that we need is
- figure out endianness
- read declared number of extended text headers
- if it is -1, calculate number of stanzas
- proceed with finding file metadata and stanzas

This breaks our attempt to combine all metadata retrievals in one C
function, as we don't read stanzas in C.

Until it is decided upon better design, explicitly set endianness in
segyopen before all other operations.

achaikou committed:
Calculate delay time in segy.c

Both DelayRecordingTime and ScalarTraceHeader are used in open.
With custom mapping those two can have any type. In other internal
functions used in open we restrict the type in segy.c so that we have a
correct state.

So it makes sense to move calculations to C, so that both tightly
coupled dt and delay were read and processes nearby.

achaikou committed:
Test opening minimal file

achaikou committed:
Use mapping in field_forall/field_foreach

Functionality is already exposed to python but is not exposed nicely to
the user.

achaikou committed:
Extract field byte-swapping code to own function

Using different function for byte-swapping may cause slight performance
degradation in segy_field_forall function, but it will likely be small.
As a benefit, new function deals automatically with byte-swapping
floats.

achaikou committed:
Support missing datatypes in core functions

achaikou committed:
Assure reading and writing fields of all types

achaikou committed:
Assure correct mapping is used in bindings

ajaust commented on lib/src/segy.c:
```diff
@@ -1846,6 +1846,84 @@ int segy_traces( segy_datasource* ds,
 }
 
+/* Allows only int2 and uint2. */
+static int field_data_to_scalar(
```
Do we expect this method to become more generalised? If not, should the restriction (only be usable for 2 byte ints) be reflected in the method's name?

achaikou replied:
No, I think this method is generalized enough.
No, I don't think restriction should be in the method's name. There are 4 scalars present in the specification, all 2 bytes.
PO decided that we should accept scalars only of this size.
Changing the name would imply other options are possible.

If someone complains in the future - then we shall look at it again.

ajaust commented on lib/src/segy.c:
```diff
@@ -1846,6 +1846,84 @@ int segy_traces( segy_datasource* ds,
+            delay_raw = delay_fd.value.u16;
+            break;
+        default:
+            return SEGY_INVALID_FIELD_DATATYPE;
```
Does it make sense to allow for `SEGY_ENTRY_TYPE_TIME2` such that we could call `field_data_to_scalar` here?

I am mainly wondering if we would have a good way to reuse `field_data_to_scalar` here instead of reimplementing (almost) the same method here.

ajaust replied:
Yeah, what I wrote is confusing.

The implementation of `field_data_to_scalar` looks almost like this `switch` statement here. `field_data_to_scalar` is only missing the `SEGY_ENTRY_TYPE_TIME2 ` case. I was wondering if it would make sense to add the `SEGY_ENTRY_TYPE_TIME2` case to `field_data_to_scalar` such that we could call `field_data_to_scalar`.

ajaust commented on lib/src/segy.c:
```diff
@@ -1846,6 +1846,84 @@ int segy_traces( segy_datasource* ds,
+        delay_scalar = -1.0f / scalar_raw;
+    } else {
+        delay_scalar = 1.0f * scalar_raw;
+    }
```
We probably don't need it here yet, but maybe this could become its own method. I assume that is the reconstruction of a `time2` value. In this sense, it could be reused other places later.

ajaust commented on python/segyio/open.py:
```diff
@@ -275,13 +275,11 @@ def _open(datasource_descriptor,
+        try:
+            t0 = fd.getdelay()
+        except RuntimeError:
+            t0 = 0.0
```
Doesn't the change modify the behavior by catching the runtime error and setting `t0` to `0` compared to before?

yngve793 commented on lib/src/segy.c:
```diff
@@ -1097,6 +1097,7 @@ static int get_field( const char* header,
             memcpy( &(fd->value.i8), header + offset, vsize );
             return SEGY_OK;
 
+
```
Unnecessary newline.

achaikou replied:
I added them on purpose to separate signed int - unsigned int - all other stuff, but will remove :+1:  .

ajaust commented on lib/src/segy.c:
```diff
@@ -1389,39 +1389,35 @@ static int slicelength( int start, int stop, int step ) {
+/* Byte-swaps field value in the header. Returns new offset after swapping or -1
+ * if error occurred.
+ */
+static int bswap_header_field_value(
```
Nice! I like this. :)

ajaust commented on test-data/README.md:
```diff
@@ -96,6 +96,8 @@ Files are created manually.
 | File                           | Purpose                                                       |
 |--------------------------------|---------------------------------------------------------------|
+| mapping-all-types.sgy          | Entries are given all types possible.                         |
+| mapping-all-types-lsb.sgy      | `flip-endianness.py mapping-all-types.sgy out.sgy`            |
```
Why not use the `flip-endianness` application in `applications/`?

ajaust commented on python/test/segyio_c.py:
```diff
@@ -443,30 +443,69 @@ def mkempty():
+        standard_index: [0x11111111, 0x44444444],
+        extension1_index: [0x22222222_22222222, 0x55555555_55555555],
+        proprietary_index: [0x3333, 0x6666],
+    }
```
We could consider making the [test parametrized](https://docs.pytest.org/en/7.1.x/example/parametrize.html) and having the `header_index` and the expected output as paremeters.

ajaust commented on python/test/segy.py:
```diff
@@ -2274,3 +2280,133 @@ def test_trace_header_extensions():
+    types_headers = {
+        0: types_header0_cases,
+        1: types_header1_cases,
+    }
```
This refers to the (extended?) trace header (=`tracehader_index = 1`) for two traces (`trace[0]` and `trace[1]`). Is this correct?

From the naming, I was initially expecting that we would read the standard trace header (index `0`) and the first extended trace header (index `1`).

yngve793 commented on python/segyio/segyio.cpp:
```diff
@@ -1132,32 +1142,318 @@ PyObject* putth( segyfd* self, PyObject* args ) {
-                                          &stop,
-                                          &step,
-                                          &field ) )
+    if( !PyArg_ParseTuple(
```
More logical indent. 👍

ajaust reviewed (commented):
I think the changes look good overall. I added some remarks about naming. Just have a look how you feel about it. Otherwise, I added some questions where I got lost and could not figure out what is going on. 🙂

yngve793 reviewed (commented):
Setting it to `Comment` for now. I have no critical comments, but it would be good to leave it for a couple of days before having a final look.

achaikou commented on python/segyio/field.py:
```diff
@@ -374,10 +374,12 @@ def __getitem__(self, key):
         """
 
-        try: return self.getfield(self.buf, int(key))
+        traceceheader_no = 0
```
0 is the only traceheader that we would refer to by name, so it won't be an enum, but a single named constant.
So I don't think it is better than named variable :woman_shrugging: 

Also what makes sense in python code can be determined only after we figure out the best design for new python features and see if those functions would be reused or not...
Maybe we can discuss again then, with more knowledge, if it still bothers us?

ajaust replied:
Yes, let's revisit later if it still bothers us then. 👍

achaikou commented on python/test/segyio_c.py:
```diff
@@ -443,30 +443,69 @@ def mkempty():
+        f.field_forall(attrs, nonexistent_index, start, stop, step, field)
+
+    with pytest.raises(KeyError):
+        f.field_foreach(attrs, nonexistent_index, indices, field)
```
I am fine with this :+1:

achaikou commented on python/test/segy.py:
```diff
@@ -2274,3 +2280,133 @@ def test_trace_header_extensions():
+            for case in header_cases:
+                offset = types_layout.entry_by_name(case.name).byte
+
+                # assert f.traceheader[trace_index][traceheader_index][offset] == case.opened
```
Yes.
We are not supposed to be using internals, those are difficult to read. But we don't have python interface in place yet, so I just wanted to be helpful and show what internals stand for :)

achaikou reviewed (commented):
Thanks for the fast reviews!! :heart_eyes_cat:

yngve793 commented on python/segyio/field.py:
```diff
@@ -374,10 +374,12 @@ def __getitem__(self, key):
         """
 
-        try: return self.getfield(self.buf, int(key))
+        traceceheader_no = 0
```
Fine

yngve793 commented on lib/src/segy.c:
```diff
@@ -1388,42 +1435,71 @@ static int slicelength( int start, int stop, int step ) {
+ */
+static int bswap_header_field_value(
+    const segy_entry_definition* mapping,
+    char* xs,
```
So, it is basically the definition of a bad variable name. :-)

achaikou replied:
Well, in our case it is an alternative to `buffer`.
I will change it if you want to!

I guess I just made my peace with it :smile: Knowing that in some languages this is a [known convention](https://stackoverflow.com/questions/6267735/what-is-the-history-of-the-variable-names-x-and-xs) also helps :smile:

ajaust commented on lib/src/segy.c:
```diff
@@ -1424,6 +1424,8 @@ static int bswap_header_word( segy_field_data* fd ) {
 
 int segy_field_forall( segy_datasource* ds,
+                       int traceheader_index,
+                       const segy_entry_definition* offset_map,
```
Ahh, thanks for the explanation. 🙂

ajaust commented on python/segyio/segyio.cpp:
```diff
@@ -2668,6 +2688,41 @@ int set_mapping_offset_to_entry_defintion(
+        if( !exp_name_heap ) return SEGY_MEMORY_ERROR;
+        std::strcpy( exp_name_heap, exp_name.c_str() );
+
+        delete[] spec_entry_name_heap;
```
Great 👍 

Regarding "objects we plan to hand over to Python" I might have confused myself. I meant these parts

```c++
segy_entry_definition def_mant = {
       SEGY_ENTRY_TYPE_SCALE6_MANT,
       requires_nonzero_value,
       mant_name_ptr.release()
};
```

where the allocated memory and the responsibility is given to someone else (the someone is probably not Python).

ajaust reviewed (approved):
It looks good to me overall. I resolved the discussions where suitable and added some replies to the ones. 

There were two or three discussions where I tried to clarify what I was trying to say. I left these discussions as unresolved so it is easier to find my reply. None of the open discussions require any further changes. 

The only open discussion that is still on my mind is the naming and description of the new test file. However, I have no immediate ideas that sound better so there is no change required. If we can come up with an improvement before we merge this PR, then that would be great. If cannot improve it, that is fine as well. I think the test file has shown its usefulness so it should be added.

yngve793 commented on python/segyio/segyio.cpp:
```diff
@@ -652,11 +652,14 @@ PyObject* segyopen( segyfd* self, PyObject* args, PyObject* kwargs ) {
-    int err = segy_binheader( ds, binary );
-    if( err ) return Error( err );
+    // endianness must be set before any other header values are read
+    if( endianness != SEGY_LSB && endianness != SEGY_MSB ) {
```
Defining SEGY_UNDIFINED=-1 in directly in SEGY_ENDIANNESS is better han hardcoding it to -1 in `segyopen` without any context. 

I expect that it already existed. Consider it.

achaikou replied:
I am not sure I agree with it conceptually and I think working on this now would take relatively too much time.
- same refers to `encoding`.
- choice between SEGY_UNDEFINED and SEGY_ENDIANNESS_UNDEFINED name, which is not 100% straightforward.
- Python code must be updated too. Python user might call `segyopen(endianness=666)`  so then if we want to have here code `if endianness != SEGY_UNDEFINED` we have to assure that user-provided endianness is inside 3 required values or code would be incorrect.

So as this is a conceptual change from local invalid value to global invalid value I would prefer to leave the investigations to a better time...

yngve793 reviewed (approved):
Added some new questions/comments.

achaikou committed:
Move getfield and setfield to new location

achaikou committed:
Use traceheaders mapping in get/set field

Get-set field would now read correct traceheader.

Note that this solution could likely be improved.

Proposed suggestion for improvement is to create a format string during
read of traceheader inside segyio.cpp (custom, python.struct, numpy's
own, doesn't matter). Then we won't need to return to C/C++ code to
parse the fields.

There might be a hurdle though: big endian <-> host endianness functions
used in get/set field. Endianness conversions are C core library
operations and are not expected to be exposed.

As this approach is not investigated yet, we keep using
getfield/setfield as before.
